paasta_tools/long_running_service_tools.py

```diff
@@ -345,20 +360,47 @@ def limit_instance_count(self, instances: int) -> int:
         return max(self.get_min_instances(), min(self.get_max_instances(), instances))
 
     def get_autoscaling_params(self) -> AutoscalingParamsDict:
-        default_params: AutoscalingParamsDict = {
-            "metrics_provider": "cpu",
+        default_provider_params: MetricsProviderDict = {
+            "type": MetricsProviderType.CPU,
             "decision_policy": "proportional",
             "setpoint": DEFAULT_AUTOSCALING_SETPOINT,
         }
-        return deep_merge_dictionaries(
-            overrides=self.config_dict.get("autoscaling", AutoscalingParamsDict({})),
-            defaults=default_params,
+
+        params = copy.deepcopy(
+            self.config_dict.get("autoscaling", AutoscalingParamsDict({}))
         )
+        if "metrics_providers" not in params or len(params["metrics_providers"]) == 0:
+            params["metrics_providers"] = [default_provider_params]
+        else:
+            for i in range(0, len(params["metrics_providers"])):
+                params["metrics_providers"][i] = deep_merge_dictionaries(
+                    overrides=params["metrics_providers"][i],
+                    defaults=default_provider_params,
+                )
+        return params
 
     def get_autoscaling_max_instances_alert_threshold(self) -> float:
         autoscaling_params = self.get_autoscaling_params()
         return autoscaling_params.get(
-            "max_instances_alert_threshold", autoscaling_params["setpoint"]
+            # TODO this default doesn't make sense for metrics providers that don't use setpoint
+            "max_instances_alert_threshold",
+            DEFAULT_AUTOSCALING_SETPOINT,
+        )
+
+    def get_autoscaling_metrics_provider(
+        self, provider_type: MetricsProviderType
+    ) -> Optional[MetricsProviderDict]:
+        autoscaling_params = self.get_autoscaling_params()
+        # We only allow one metric provider of each type, so we can bail early if we find a match
+        for provider in autoscaling_params["metrics_providers"]:
+            if provider["type"] == provider_type:
+                return provider
+        return None
+
+    def should_use_metrics_provider(self, provider_type: MetricsProviderType) -> bool:
```

Comment:
comment: i'd say we could lru_cache this - but it doesn't seem like this is getting called in a hot loop anywhere so probably fine

Reply:
(and also ++ to this refactor)

Reply:
eh, YAGNI, probably